descartes_moveit/src/seed_search.cpp

```diff
@@ -0,0 +1,455 @@
+  return true;
+}
+
+/**
+ * @brief Returns true if the determinant of the jacobian is not zero.
+ *        Returns false if it is close to zero (and robot is near singularity)
+ */
+bool checkSingularity(moveit::core::RobotState& state,
+                      const moveit::core::JointModelGroup* group)
+{
+ Eigen::Vector3d reference_point_position(0.0, 0.0, 0.0);
+ Eigen::MatrixXd jacobian;
+ state.getJacobian(group, state.getLinkModel(group->getLinkModelNames().back()),
+                   reference_point_position, jacobian);
+
+ return std::abs(jacobian.determinant()) > 0.0001;
+}
+
+// Compares the first n_compare bijection joint values from a and b and tests
+// whether they are within 45 degrees of one another
+inline bool sameJointConfig(const seed::JointConfig& a, const seed::JointConfig& b,
+                            const BijectionVec& bij, unsigned n_compare)
+{
+  for (std::size_t i = 0; i < n_compare; i++)
+  {
+    if (std::fmod(std::abs(a[bij[i]] - b[bij[i]]), 2*M_PI) > M_PI_4) return false;
+  }
+  return true;
+}
+
+// Tests whether a given joint configuration, c, is similar to any joint set already
+// in set using the given bijection values and number of joints.
+inline bool inJointSet(const seed::JointConfig& c, const seed::JointConfigVec& set,
+                       const BijectionVec& bij, unsigned n_compare)
+{
+  for (std::size_t i = 0; i < set.size(); ++i)
+  {
+    if (sameJointConfig(c, set[i], bij, n_compare)) return true;
+  }
+  return false;
+}
+
+// Creates a bijection vector which puts the permuted pair of joints
+// up front.
+BijectionVec createBijection(const seed::JointPair& pair, unsigned dof)
+{
+  BijectionVec bij;
+  bij.reserve(dof);
+
+  // add first two joints
+  bij.push_back(pair.first);
+  bij.push_back(pair.second);
+
+  for (std::size_t i = 0; i < dof; i++)
+  {
+    if (i != pair.first && i != pair.second)
+      bij.push_back(i);
+  }
+
+  return bij;
+}
+
+// Uses ROS's random number generator to fill out a vector of doubles (length dof)
+// with uniform values between min and max
+std::vector<double> randomInitialState(unsigned dof, double min, double max)
+{
+  std::vector<double> result;
+  result.reserve(dof);
+
+  random_numbers::RandomNumberGenerator gen;
+
+  for (std::size_t i = 0; i < dof; ++i)
+  {
+    result.push_back(gen.uniformReal(min, max));
+  }
+  return result;
+}
+
+/**
+ * @brief Returns possible seed states generated from permuting the given pair of joints
+ *        (using joint numbering starting at 0)
+ */
+seed::JointConfigVec findSeedStatesForPair(moveit::core::RobotStatePtr state,
+                                           const std::string& group_name,
+                                           const std::string& tool_frame,
+                                           const seed::JointPair& pair)
+{
+  const moveit::core::JointModelGroup* group = state->getJointModelGroup(group_name);
+
+  // Hard-coded constants for joint iteration
```

Comment:
Well this begs a couple of questions:
1. Should we handle joint types that aren't revolute?
2. How should we discretize the valid joint space? Should I walk 90 degrees or the end of the range, which ever comes first?